presto-main/src/main/java/com/facebook/presto/operator/SpatialIndexBuilderOperator.java

```diff
@@ -0,0 +1,231 @@
+        {
+            throw new UnsupportedOperationException("Spatial index build can not be duplicated");
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final LocalMemoryContext localUserMemoryContext;
+    private final PagesSpatialIndexFactory pagesSpatialIndexFactory;
+
+    private final List<Integer> outputChannels;
+    private final int indexChannel;
+    private final BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest;
+    private final Optional<JoinFilterFunctionFactory> filterFunctionFactory;
+
+    private final PagesIndex index;
+    private ListenableFuture<?> indexNotNeeded;
+
+    private boolean finished;
+
+    public SpatialIndexBuilderOperator(
+            OperatorContext operatorContext,
+            PagesSpatialIndexFactory pagesSpatialIndexFactory,
+            List<Integer> outputChannels,
+            int indexChannel,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory,
+            int expectedPositions,
+            PagesIndex.Factory pagesIndexFactory)
+    {
+        this.operatorContext = operatorContext;
+        this.localUserMemoryContext = operatorContext.localUserMemoryContext();
+        this.spatialRelationshipTest = spatialRelationshipTest;
+        this.filterFunctionFactory = filterFunctionFactory;
+
+        this.index = pagesIndexFactory.newPagesIndex(pagesSpatialIndexFactory.getTypes(), expectedPositions);
+        this.pagesSpatialIndexFactory = pagesSpatialIndexFactory;
+
+        this.outputChannels = outputChannels;
+        this.indexChannel = indexChannel;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return pagesSpatialIndexFactory.getTypes();
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+        checkState(!isFinished(), "Operator is already finished");
+
+        index.addPage(page);
+
+        if (!localUserMemoryContext.trySetBytes((index.getEstimatedSize().toBytes()))) {
+            index.compact();
+            localUserMemoryContext.setBytes(index.getEstimatedSize().toBytes());
+        }
+
+        operatorContext.recordGeneratedOutput(page.getSizeInBytes(), page.getPositionCount());
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        return null;
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        if (indexNotNeeded != null && !indexNotNeeded.isDone()) {
+            return indexNotNeeded;
+        }
+        return NOT_BLOCKED;
+    }
+
+    @Override
+    public void finish()
+    {
+        if (finished) {
+            return;
+        }
+
+        if (indexNotNeeded == null) {
```

Comment:
I thought for a while  `indexNotNeeded != null` is not possible... I would suggest using a boolean for state-control rather than nullness of `indexNotNeeded`:

- `finish()` to set boolean `finished=true` + rename to `finishing`:
```
if (finishing) {
   return;
}

finishing = true;
... create spatial index
```

- `isFinished` to consult both `finishing` and `indexNotNeeded.isDone()`:
```
if (!finishing) {
  return false;
}

verify(indexNotNeeded != null);
return indexNotNeeded.isDone(); // + the memory cleanup you do before returning true ... so you might need separate `finished` boolean still
```